internal/extender/resource.go

```diff
@@ -294,11 +292,73 @@ func (s *SparkSchedulerExtender) selectDriverNode(ctx context.Context, driver *v
 	return reservedDriverNode, outcome, err
 }
 
-func (s *SparkSchedulerExtender) potentialNodes(availableNodes []*v1.Node, driver *v1.Pod, nodeNames []string) (driverNodes, executorNodes []string) {
-	sort.Slice(availableNodes, func(i, j int) bool {
-		return availableNodes[j].CreationTimestamp.Before(&availableNodes[i].CreationTimestamp)
+const olderNodesThreshold = time.Duration(-2 * 3600 * 1000 * 1000 * 1000)
+
+// Sort older nodes before younger nodes.
+// Sort older nodes by node age ascending.
+// Sort younger nodes by available resources ascending, with RAM usage more important.
+func compareNodes(left scheduleContext, right scheduleContext, now time.Time) bool {
+	var threshold = now.Add(olderNodesThreshold)
+	if left.creationTime.Before(threshold) && right.creationTime.Before(threshold) {
+		return left.creationTime.After(right.creationTime)
+	} else if left.creationTime.Before(threshold) {
+		return true
+	} else if right.creationTime.Before(threshold) {
+		return false
+	} else if left.availableMemory.Cmp(right.availableMemory) != 0 {
+		return left.availableMemory.Cmp(right.availableMemory) == -1
+	} else if left.availableCPU.Cmp(right.availableCPU) != 0 {
+		return left.availableCPU.Cmp(right.availableCPU) == -1
+	}
+	return left.creationTime.Before(right.creationTime)
+}
+
+type scheduleContext struct {
+	creationTime    time.Time
+	availableMemory resource.Quantity
+	availableCPU    resource.Quantity
+}
+
+func cpu(ctx context.Context, r *resources.Resources) resource.Quantity {
```

Comment:
these methods are a bit suspect. Can I guarantee that the result of s.usedResources definitely contains all of the list of nodes I've been handed?